elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,694 @@
+    final List<Hypercube> H = getSortedHypercubes(relation);
+
+    LOG.beginStep(stepprog, 2, "Obtaining densities per hypercube.");
+    final int[] W = this.strategy.buildIndex(H).getDensities();
+
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+    {
+      // compute score by hypercube
+      LOG.beginStep(stepprog, 3, "Computing hypercube scores");
+
+      FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("HySortOD scores", relation.size(), LOG) : null;
+      DBIDArrayIter iter = (DBIDArrayIter) relation.iterDBIDs();
+      for(int hypercube = 0; hypercube < H.size(); hypercube++) {
+        final IntegerArray instances = H.get(hypercube).getInstances();
+        final double hypercubeScore = score(W[hypercube]);
+        minmax.put(hypercubeScore);
+        for(int instance = 0; instance < instances.size; instance++) {
+          final int index = instances.get(instance);
+          scores.putDouble(iter.seek(index), hypercubeScore);
+          LOG.incrementProcessed(prog);
+        }
+      }
+      LOG.ensureCompleted(prog);
+    }
+
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("HySortOD", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  /**
+   * Create and sort hypercubes considering their coordinates.
+   * 
+   * @param relation Data to process
+   * @return Hypercubes sorted
+   */
+  private List<Hypercube> getSortedHypercubes(Relation<V> relation) {
+    TreeSet<Hypercube> sorted = new TreeSet<HySortOD<V>.Hypercube>();
+
+    // Iterate over all objects
+    for(DBIDArrayIter iter = (DBIDArrayIter) relation.iterDBIDs(); iter.valid(); iter.advance()) {
+      Hypercube h = new Hypercube(relation.get(iter).toArray(), this.l);
+      if(!sorted.contains(h)) {
+        h.add(iter.getOffset());
+        sorted.add(h);
+      }
+      else {
+        sorted.ceiling(h).add(iter.getOffset());
+      }
+    }
+
+    int n = sorted.size();
+    List<Hypercube> H = new ArrayList<Hypercube>(n);
+
+    for(Hypercube h : sorted) {
+      H.add(h);
+    }
+
+    return H;
+  }
+
+  /**
+   * Compute score according to hypercube neighborhood density.
+   * 
+   * @param density Hypercube density
+   * @return score
+   */
+  private double score(int density) {
+    return 1 - (density / this.strategy.getMaxDensity());
+  }
+
+  @Override
+  public TypeInformation[] getInputTypeRestriction() {
+    return TypeUtil.array(TypeUtil.NUMBER_VECTOR_FIELD);
+  }
+
+  /**
+   * Bounded regions of the space where at least one instance exists.
+   * 
+   * @author Cabral, Eugênio F.
+   */
+  class Hypercube implements Comparable<Hypercube> {
```

Comment:
The class can be made `static`, which makes the objects smaller (no pointer to parent).